File: Editorial/0130_Surrounded_Regions/Go/2_BFS.go
type pair struct {
    first  int
    second int
}

func solve(board [][]byte) {
    if len(board) == 0 || len(board[0]) == 0 {
        return
    }
    ROWS := len(board)
    COLS := len(board[0])
    border := make([]pair, 0)
    for r := 0; r < ROWS; r++ {
        border = append(border, pair{r, 0})
        border = append(border, pair{r, COLS - 1})
    }
    for c := 0; c < COLS; c++ {
        border = append(border, pair{0, c})
        border = append(border, pair{ROWS - 1, c})
    }
    for _, pair := range border {
        BFS(board, pair.first, pair.second, ROWS, COLS)
    }
    for r := 0; r < ROWS; r++ {
        for c := 0; c < COLS; c++ {
            if board[r][c] == 'O' {
                board[r][c] = 'X'
            }
            if board[r][c] == 'E' {
                board[r][c] = 'O'
            }
        }
    }
}

func BFS(board [][]byte, row int, col int, ROWS int, COLS int) {
    queue := make([]pair, 0)
    queue = append(queue, pair{row, col})
    for len(queue) != 0 {
        top := queue[0]
        queue = queue[1:]
        row = top.first
        col = top.second
        if board[row][col] != 'O' {
            continue
        }
        board[row][col] = 'E'
        if col < COLS-1 {
            queue = append(queue, pair{row, col + 1})
        }
        if row < ROWS-1 {
            queue = append(queue, pair{row + 1, col})
        }
        if col > 0 {
            queue = append(queue, pair{row, col - 1})
        }
        if row > 0 {
            queue = append(queue, pair{row - 1, col})
        }
    }
}

File: Editorial/0101-0200/0130_Surrounded_Regions/TypeScript/2_BFS.ts
var solve = function (board: Array<Array<string>>): void {
    if (!board || !board[0]) {
        return;
    }
    const ROWS: number = board.length;
    const COLS: number = board[0].length;
    let borders: number[][] = Array.from({ length: ROWS }, (v, i) => i)
        .map((i) => [
            [i, 0],
            [i, COLS - 1],
        ])
        .concat(
            Array.from({ length: COLS }, (v, i) => i).map((i) => [
                [0, i],
                [ROWS - 1, i],
            ]),
        )
        .flat();
    borders.forEach(([row, col]) => BFS(board, row, col, ROWS, COLS));
    for (let row = 0; row < ROWS; ++row)
        for (let col = 0; col < COLS; ++col) {
            if (board[row][col] == "O") board[row][col] = "X";
            if (board[row][col] == "E") board[row][col] = "O";
        }
};
var BFS = function (
    board: Array<Array<string>>,
    row: number,
    col: number,
    ROWS: number,
    COLS: number,
): void {
    let queue: number[][] = [[row, col]];
    while (queue.length) {
        let [row, col] = queue.shift() as number[];
        if (board[row][col] !== "O") {
            continue;
        }
        board[row][col] = "E"; // Mark this cell as escaped
        if (col < COLS - 1) queue.push([row, col + 1]);
        if (row < ROWS - 1) queue.push([row + 1, col]);
        if (col > 0) queue.push([row, col - 1]);
        if (row > 0) queue.push([row - 1, col]);
    }
};
